nautobot/extras/models/groups.py

```diff
@@ -891,83 +731,80 @@ def perform_membership_set_operation(self, operator, query, next_set):
-        if not memberships.exists():
-            return self.generate_members_query()
+        if self.group_type == DynamicGroupTypeChoices.TYPE_DYNAMIC_SET:
+            query = models.Q()
+            memberships = self.dynamic_group_memberships.all()
+            # Enumerate the filters for each child group, trusting that they handle their own children.
+            for membership in memberships:
+                group = membership.group
+                operator = membership.operator
+                logger.debug("Processing group %s...", group)
 
-        # Enumerate the filters for each child group, trusting that they handle their own children.
-        for membership in memberships:
-            group = membership.group
-            operator = membership.operator
-            logger.debug("Processing group %s...", group)
+                if group.group_type == DynamicGroupTypeChoices.TYPE_DYNAMIC_FILTER:
+                    logger.debug("Query: %s -> %s -> %s", group, group.filter, operator)
+
+                next_set = group.generate_query()
+                query = self._perform_membership_set_operation(operator, query, next_set)
 
-            if group.filter:
-                logger.debug("Query: %s -> %s -> %s", group, group.filter, operator)
+            return query
 
-            next_set = group.generate_members_query()
-            query = self.perform_membership_set_operation(operator, query, next_set)
+        # TODO? if self.group_type == DynamicGroupTypeChoices.TYPE_STATIC:
 
-        return query
+        raise RuntimeError(f"generate_query not implemented for group_type {self.group_type}")
 
-    def get_group_queryset(self):
-        """Return a filtered queryset of all descendant groups."""
+    def _get_group_queryset(self):
+        """Construct the queryset representing dynamic membership of this group."""
         query = self.generate_query()
-        qs = self.get_queryset()
-        return qs.filter(query)
+        return self.model.objects.filter(query)
 
+    # TODO: unused in core
     def add_child(self, child, operator, weight):
         """
         Add a child group including `operator` and `weight`.
 
-        :param child:
-            DynamicGroup instance
-        :param operator:
-            DynamicGroupOperatorChoices choice value used to dictate filtering behavior
-        :param weight:
-            Integer weight used to order filtering
+        Args:
+            child (DynamicGroup): child group to add
+            operator (str): DynamicGroupOperatorChoices choice value used to dictate filtering behavior
+            weight (int): Integer weight used to order filtering
         """
+        if self.group_type != DynamicGroupTypeChoices.TYPE_DYNAMIC_SET:
+            if self.filter or self.group_type != DynamicGroupTypeChoices.TYPE_DYNAMIC_FILTER:
+                raise ValidationError(f"{self} is not a dynamic-set group.")
```

Comment:
I am a little confused about the code here. Why is the second if statement there?

Reply:
Before this PR, the only way of distinguishing between a "filter-based" DG and a "set-based" DG was by whether it contained any other groups - if it contained them, it was set-based (even if it had a defined `filter` otherwise) and if it didn't, it was filter-based (even if that filter was empty).  Now that we have explicit `group_type` values, I wanted to remain ORM-backward-compatible with the old pattern where every group started off as filter-based, then upon adding its first child group, became set-based.

So:

- if calling add_child and we're not currently set-based:
    - if we already have an explicit filter (`self.filter` is truthy), we can't become set-based
    - if we're a static group (or anything other than filter-based) we also can't become set-based
    - else, "magically" become set-based at this point.